### PR TITLE
feat: introduce logic for distillation validation component

### PR DESCRIPTION
**Pull request by ShobithNandakumar**
### Description

This PR introduces the core logic for the distillation validation component while the parent PR holds other changes. 

Reason for change will be documented in the parent PR. <parent PR to be updated here>

#### Tests Conducted

data_generation_task_type | enable_chain_of_thought | input_data | Status | Output (truncated)
-- | -- | -- | -- | --
CONVERSATION | True | Quora/alpaca | Fail | `UserError - Error validating dataset record 0: Chain of thought is not supported for task type CONVERSATION`
NLI/NLU_QA | True | cestwc/conjnli tau/commonsense_qa | Success | `validation_status: ok`
CONVERSATION | False | Quora/alpaca | Success | `validation_status: ok`
CONVERSATION | False | cestwc/conjnli | Fail | `UserError - Error validating dataset record 0: Dataset needs to be of multi-turn for task type CONVERSATION`
NLI/NLU_QA | True/False | Quora/alpaca | Fail | `UserError - Error validating dataset record 0: Chat cannot be of type multi-turn for task type NLI/NLU_QA`

### CODE REVIEW

File: assets/training/distillation/src/validate_pipeline.py [hunk 1]
```diff
@@ -3,33 +3,335 @@
...
+            MAX_NEW_TOKENS: self._args.teacher_model_max_new_tokens,
+            TEMPERATURE: self._args.teacher_model_temperature,
+            TOP_P: self._args.teacher_model_top_p,
+            **record
+        }
+
+        headers = self._get_inference_request_headers()
+        url = self._args.teacher_model_endpoint_url
+        response = requests.post(url=url, headers=headers, data=json.dumps(inference_params))
+        response.raise_for_status()
```
Comment (novaturient95) [inline assets/training/distillation/src/validate_pipeline.py:138]: could you please raise UserError here

Comment (ShobithNandakumar) [reply assets/training/distillation/src/validate_pipeline.py:140]: Thanks @novaturient95 for catch this! I'm supporting exponential backoff on failures. Please have a look now.

**Review by novaturient95 — APPROVED** [review]

**Comment by babu-namburi** [comment]
Can you please include the below test cases when you are testing
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>


<div style='direction:ltr;border-width:100%'>

<div style='direction:ltr;margin-top:0in;margin-left:0in;width:5.877in'>

<div style='direction:ltr;margin-top:0in;margin-left:0in;width:5.877in'>

<div style='direction:ltr'>


data_generation_task_type | enable_chain_of_thought | input_data | Status
-- | -- | -- | --
CONVERSATION | True | Quora/alpaca | Fail
NLI/NLU_QA | True | cestwc/conjnli    tau/commonsense_qa | Success
CONVERSATION | False | Quora/alpaca | Success
CONVERSATION | False | cestwc/conjnli | Fail
NLI/NLU_QA | True/False | Quora/alpaca | Fail
</div>

</div>

</div>

</div>


</body>
</html>

**Comment by babu-namburi** [comment]
@ShobithNandakumar 
Also, can you please check if the environment used in distillation has this https://pypi.org/project/backoff/ 
So that we don't have to maintain the backoff code.

**Comment by ShobithNandakumar** [comment]
> @ShobithNandakumar Also, can you please check if the environment used in distillation has this https://pypi.org/project/backoff/ So that we don't have to maintain the backoff code.

@babu-namburi - yep had verified it was not in use before I added a custom decorator for this :) 
Packages are listed [here](https://ml.azure.com/registries/azureml/environments/acft-hf-nlp-gpu/version/63?tid=7f292395-a08f-4cc0-b3d0-a400b023b0d2)

**Comment by novaturient95** [comment]
for exponential backoff, is urllib3 retry a valid option? @babu-namburi @ShobithNandakumar 

```python


from urllib3.util import Retry
from requests import Session
from requests.adapters import HTTPAdapter

s = Session()
retries = Retry(
    total=3,
    backoff_factor=0.1,
    status_forcelist=[502, 503, 504],
    allowed_methods={'POST'},
)
s.mount('https://', HTTPAdapter(max_retries=retries))

```

**Comment by babu-namburi** [comment]
> for exponential backoff, is urllib3 retry a valid option? @babu-namburi @ShobithNandakumar
> 
> ```python
> from urllib3.util import Retry
> from requests import Session
> from requests.adapters import HTTPAdapter
> 
> s = Session()
> retries = Retry(
>     total=3,
>     backoff_factor=0.1,
>     status_forcelist=[502, 503, 504],
>     allowed_methods={'POST'},
> )
> s.mount('https://', HTTPAdapter(max_retries=retries))
> ```

I didn't know that urllib3 has this nice functionality, 
Is urllib3 part of our environment??  I dont see it in the requirements.txt of the environment is it getting installed as part of transitive dependencies?

**Comment by ShobithNandakumar** [comment]
> for exponential backoff, is urllib3 retry a valid option? @babu-namburi @ShobithNandakumar
> 
> ```python
> from urllib3.util import Retry
> from requests import Session
> from requests.adapters import HTTPAdapter
> 
> s = Session()
> retries = Retry(
>     total=3,
>     backoff_factor=0.1,
>     status_forcelist=[502, 503, 504],
>     allowed_methods={'POST'},
> )
> s.mount('https://', HTTPAdapter(max_retries=retries))
> ```

Let me give it a try.

**Pull request merged**